core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -307,11 +322,18 @@ class BrokerMetadataListener(
 
   private def publish(publisher: MetadataPublisher): Unit = {
     val delta = _delta
-    _image = _delta.apply()
+    try {
+      _image = _delta.apply()
+    } catch {
+      case t: Throwable => metadataLoadingFaultHandler.handleFault(s"Error applying metadata delta $delta", t)
+    }
+
     _delta = new MetadataDelta(_image)
     if (isDebugEnabled) {
       debug(s"Publishing new metadata delta $delta at offset ${_image.highestOffsetAndEpoch().offset}.")
     }
+
+    // This publish call is done with its own try-catch and fault handler
     publisher.publish(delta, _image)
```

Comment:
Is this correct? If there was an error in `_image = _delta.apply()`, `_image` will be the previous image that was published while `delta` is the new `_delta` that was not applied. Also, note that I am pretty sure that the code `publisher.publish` assumes that this layer doesn't send duplicate deltas and images.

Is there a way we can write tests for this code and scenarios so that we can increase our confidence that this code behaves as we expect?

Reply:
Thanks, good catch. I missed a `throw` in the catch block above. If we can't apply the delta we should not publish the image.